core/src/main/java/org/pentaho/di/metastore/MetaStoreConst.java

```diff
@@ -66,38 +70,128 @@ public class MetaStoreConst {
+  // a Supplier than a specific instance.
+
   private static final Supplier<MetastoreLocator> metastoreLocatorSupplier = new Supplier<MetastoreLocator> () {
     private volatile MetastoreLocator metastoreLocator;
 
     public MetastoreLocator get() {
-      if ( metastoreLocator == null ) {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreLocator result = metastoreLocator;
+      if ( result == null ) {
         synchronized ( this ) {
           try {
             if ( metastoreLocator == null ) {
               Collection<MetastoreLocator> metastoreLocators =
                 PluginServiceLoader.loadServices( MetastoreLocator.class );
-              metastoreLocator = metastoreLocators.stream().findFirst().orElse( null );
+              metastoreLocator = result = metastoreLocators.stream().findFirst().orElse( null );
             }
           } catch ( KettlePluginException e ) {
             logger.error( "Error getting metastore locator", e );
           }
         }
       }
-      return metastoreLocator;
+      return result;
+    }
+  };
+
+  private static final MetastoreDirectoryProvider NULL_METASTORE_DIRECTORY_PROVIDER =
+    new MetastoreDirectoryProvider() {
+      @Override
+      public IMetaStore getMetastoreForDirectory(String rootFolder) {
+        return null;
+      }
+    };
+
+  private static final Supplier<MetastoreDirectoryProvider> metastoreDirectoryProviderSupplier =
+    new Supplier<MetastoreDirectoryProvider> () {
+    private volatile MetastoreDirectoryProvider metastoreDirectoryProvider;
+
+    public MetastoreDirectoryProvider get() {
+      // double-checked idiom because this can fail until the plugin is loaded.
+      MetastoreDirectoryProvider result = metastoreDirectoryProvider;
+      if ( result == null ) {
+        synchronized ( this ) {
+          try {
+            if ( metastoreDirectoryProvider == null ) {
+              Collection<MetastoreDirectoryProvider> metastoreDirectoryProviders =
+                PluginServiceLoader.loadServices( MetastoreDirectoryProvider.class );
+              metastoreDirectoryProvider = metastoreDirectoryProviders.stream().findFirst().orElse( null );
+              if ( metastoreDirectoryProvider == null ) {
+                // cache the null so we don't constantly try to look this up.
+                metastoreDirectoryProvider = result = NULL_METASTORE_DIRECTORY_PROVIDER;
+              }
+            }
+          } catch ( KettlePluginException e ) {
+            logger.error( "Error getting metastore directory provider", e );
+          }
+        }
+      }
+      return result;
+    }
+  };
+
+  private static final MetastoreBowlProvider NULL_METASTORE_BOWL_PROVIDER =
+    new MetastoreBowlProvider() {
+      @Override
+      public IMetaStore getExplicitMetastoreForBowl(Bowl bowl) {
+        return null;
```

Comment:
if this can return null, maybe make the interface return Optional? Or at least add to the javadoc

Reply:
The MetastoreBowlProvider interface will basically only ever be used by the Supplier getters in MetaStoreConst, so the "perfect" API is not worth making it inconsistent with our other plugin APIs. Do we have other plugin interfaces that use Optional? 

I can certainly add it to the interface javadoc. 

I wanted this case to throw an exception instead of returning null, but couldn't get KettleException to work.

Reply:
Also, as far as I can tell, these should never actually return null in practice. Nothing should be asking for an explicit bowl metastore unless the project plugin is present.